cli/source/hooks/usePreCheck.ts

```diff
@@ -0,0 +1,80 @@
+import { isFreePort } from 'find-free-ports';
+import { useState } from 'react';
+
+import fs from 'fs';
+
+import { PreCheckStates } from '../constants.js';
+import { runCommand } from '../utils/run-command.js';
+
+export const usePreCheck = ({
+  db,
+  redis,
+  frontend,
+  sync_server,
+  analytics_server
+}: {
+  db: number;
+  redis: number;
+  frontend: number;
+  sync_server: number;
+  analytics_server: number;
+}) => {
+  const [daemon, setDaemon] = useState<PreCheckStates>(PreCheckStates.RUNNING);
+  const [ports, setPorts] = useState<PreCheckStates>(PreCheckStates.RUNNING);
+  const [composeFile, setComposeFile] = useState<PreCheckStates>(
+    PreCheckStates.RUNNING
+  );
+  const [dockerFile, setDockerFile] = useState<PreCheckStates>(
+    PreCheckStates.RUNNING
+  );
+
+  const callChecks = async () => {
+    // For Docker daemon
+    runCommand('docker', ['info'])
+      .promise.then(() => {
+        setDaemon(PreCheckStates.SUCCESS);
+      })
+      .catch((err) => {
+        setDaemon(PreCheckStates.FAILED);
+      });
+
+    // For ports
+    if (
+      isNaN(db) ||
+      isNaN(redis) ||
+      isNaN(frontend) ||
+      isNaN(sync_server) ||
+      isNaN(analytics_server)
+    ) {
+      setPorts(PreCheckStates.FAILED);
+    } else {
+      const db_check = await isFreePort(db);
+      const redis_check = await isFreePort(redis);
+      const fe_check = await isFreePort(frontend);
+      const ss_check = await isFreePort(sync_server);
+      const as_check = await isFreePort(analytics_server);
+      if (!db_check || !redis_check || !fe_check || !ss_check || !as_check) {
+        setPorts(PreCheckStates.FAILED);
+      } else {
+        setPorts(PreCheckStates.SUCCESS);
+      }
+    }
```

Comment:
You can use `Promise.all()` and `Promise.allSettled()` to avoid having to store values of promises in variables and conditionally checking them. It might make code more readable, imo. Just a suggestion!

Reply:
So here what would be the better option between `Promise.all()` and `Promise.allSettled()`? If we want better display of failed checks, then it might be better to also display the exact port where it fails (unlike what is being done in the current UI that I implemented). Would the `Promise.allSettled()` be better in that case as it would run all the checks even if one of them fails?

Reply:
Definitely allSettled.
I don't know about readability, but it'll def be faster.

My suggestion:

```js
const portPromises = [db, redis, frontend, sync_server, analytics_server].map(isFreePort);
const checks = await Promise.allSettled(portPromises);
if (checks.includes(false)) failed;
else passed;
```